pkg/policy: correctly merge L4 rules which apply to the same port

ianvernon opened the pull request:
* pkg/k8s: add TODO for cleaning up unit tests
* Documentation/gettingstarted: update `cilium status` output in Docker GSG
* Documentation/gettingstarted: update Minikube GSG to reflect how we handle L4-only and L4-L7 policy on the same port
    - The GSG needs to be updated to account for the fact that when one rule selects an endpoint with L4-only policy, and another rule selects the same endpoint on the same port/protocol, we resolve internally within Cilium to allow all traffic on L4. Thus, use a CiliumNetworkPolicy with the same name that is used for L4-only policy, and then updated with L4-L7 policy.
* examples/minikube: convert L3-L4 policy to CiliumNetworkPolicy
* pkg/policy/api: add SelectsAllEndpoints function
    - This function tests whether the EndpointSelectorSlice selects all endpoints.
* pkg/policy: fix merging of L4-related policy
    -L4Filter: document what empty L7Parser represents in L4Filter.
    -  add L7ParserType to represent no L7Parser: instead of using an empty string define a constant which represents no parser type.
    - remove addEndpoints function: this function is no longer used; remove it.
    - change type of Endpoints field in L4Filter to EndpointSelectorSlice
        - This should not result in any functional change. This change was made for the L4Filter to consume `EndpointSelectorSlice.SelectsAllEndpoints()`.
    - add AllowsAllAtL3 function for L4Filter
        - This contains the logic to determine whether an L4Filter selects all endpoints at L3.
    - do not treat empty Endpoints in L4Filter as selecting all endpoints
        - For consistency across the policy package, set the slice to contain the `api.WildcardEndpointSelector`, which explicitly shows that the L4Filter applies to all endpoints. If no endpoints are provided when an L4Filter is created, set the list of endpoints to be a slice containing `api.WildcardEndpointSelector`.
    - properly handle merging ingress L4-only and L4-L7 rules which apply to the same port/protocol tuple

For instance, take the following CiliumNetworkPolicy:

```
    specs:
    - endpointSelector:
        matchLabels:
          id: app1
      ingress:
      - fromEndpoints:
        - matchLabels:
            id: app2
        toPorts:
        - ports:
          - port: "9093"
            protocol: TCP
          rules:
            http:
            - method: ^GET$
              path: ^/metrics$
      - toPorts:
        - ports:
          - port: "9093"
            protocol: TCP
```
    
These two separate rules previously resolved to only allowing traffic on `9093/TCP` on `^GET$ ^/metrics$` . This is incorrect, as the toPorts rule implicitly allows all traffic on the port, regardless of the L7 rule.
    
This commit changes how merging of L4 port rules is done to ensure that  the resultant policy allows all traffic on 9093/TCP, ignoring all L7-rules.
Signed-off by: Ian Vernon <ian@cilium.io>

Still todo:
- [x] Review the getting-started-guides to account for this, as it will be affected by this change.
    - [x] Kubernetes GSG
    - TODO (other GSGs?)
- [x] Update Ginkgo tests.
    - To my surprise, the tests passed, so I guess this doesn't need to be done :-) 
- [x] Make corresponding changes on merging egress L4 port rules.

Related-to: #3509

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3518)
<!-- Reviewable:end -->

Commit:
pkg/policy: fix merging of L4-related policy

* L4Filter: document what empty L7Parser represents in L4Filter.
* add L7ParserType to represent no L7Parser: instead of using an empty string
define a constant which represents no parser type.
* remove addEndpoints function: this function is no longer used; remove it.
* change type of Endpoints field in L4Filter to EndpointSelectorSlice
    - This should not result in any functional change. This change was made for
      the L4Filter to consume `EndpointSelectorSlice.SelectsAllEndpoints()`.
* add AllowsAllAtL3 function for L4Filter
    - This contains the logic to determine whether an L4Filter selects all
      endpoints at L3.
* do not treat empty Endpoints in L4Filter as selecting all endpoints
    - For consistency across the policy package, set the slice to contain the
      `api.WildcardEndpointSelector`, which explicitly shows that the L4Filter
      applies to all endpoints. If no endpoints are provided when an L4Filter
      is created, set the list of endpoints to be a slice containing
      `api.WildcardEndpointSelector`.
* properly handle merging ingress L4-only and L4-L7 rules which apply to the
same port/protocol tuple

Take the following CiliumNetworkPolicy:
```
specs:
- endpointSelector:
    matchLabels:
      id: app1
  ingress:
  - fromEndpoints:
    - matchLabels:
        id: app2
    toPorts:
    - ports:
      - port: "9093"
        protocol: TCP
      rules:
        http:
        - method: ^GET$
          path: ^/metrics$
  - toPorts:
    - ports:
      - port: "9093"
        protocol: TCP
```

These two separate rules previously resolved to only allowing traffic on
9093/TCP on ^GET$ ^/metrics$ . This is incorrect, as the toPorts rule
implicitly allows all traffic on the port, regardless of the L7 rule.

This commit changes how merging of L4 port rules is done to ensure that
the resultant policy allows all traffic on 9093/TCP, ignoring all L7-rules.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
examples/minikube: convert L3-L4 policy to CiliumNetworkPolicy

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
Documentation/gettingstarted: update Minikube GSG to reflect how we handle L4-only and L4-L7 policy on the same port

The GSG needs to be updated to account for the fact that when one rule selects an endpoint with L4-only policy,
and another rule selects the same endpoint on the same port/protocol, we resolve internally within Cilium
to allow all traffic on L4. Thus, use a CiliumNetworkPolicy with the same name that is used for L4-only policy,
and then updated with L4-L7 policy.

Signed-off by: Ian Vernon <ian@cilium.io>

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -88,7 +88,8 @@ type L4Filter struct {
 	Endpoints []api.EndpointSelector `json:"-"`
-	// L7Parser specifies the L7 protocol parser (optional)
+	// L7Parser specifies the L7 protocol parser (optional). If specified as
+	// an empty string, then means that no L7 forwarding is performed to the proxy.
```
L7 forwarding is performed to the proxy -> L7 proxy redirect is performed
(we've been using the term "redirect" quite consistently)

rlenglet commented on pkg/policy/repository_test.go:
```diff
@@ -415,13 +415,9 @@ func (ds *PolicyTestSuite) TestMinikubeGettingStarted(c *C) {
-			},
-		},
+		Endpoints:        selectorFromApp2DupList,
+		L7Parser:         "",
```
Define a constant for the empty string, along with the other constants of this type.

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -57,6 +57,11 @@ func (policy *L4Filter) addEndpoints(endpoints []api.EndpointSelector) bool {
 	return false
 }
 
+// mergeL4IngressPort merges all rules which share the same port & protocol that
```
Fantastic docs, thanks!

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -57,6 +57,11 @@ func (policy *L4Filter) addEndpoints(endpoints []api.EndpointSelector) bool {
+// select a given set of endpoints. It updates the L4Filter mapped to by the specified
+// port and protocol with the contents of the provided PortRule. If the rule
+// being merged has conflicting L7 rules with those already in the provided
+// L4PolicyMap for the specified port-protocol tuple, returns an error.
```
returns -> it returns

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -66,6 +71,30 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
 	}
+
+	if filter.addEndpoints(endpoints) && r.NumRules() == 0 {
+		// skip this policy as it is already covered and it does not contain L7 rules
```
I think this is incorrect.

The `addEndpoints` method only resolves which endpoints are allowed to connect at L3 and above.
When `filter.addEndpoints` returns true, it only means that connectivity at L3 was already allowed. But it provides no indication whether there are L7 rules involved, i.e. L3/label-dependent L7.

If the new rule has no L7 rules, it means that the new rule is L3+L4 only. If there were L3+L4+L7 rules associated with this L4 port and endpoint selector, those rules must be replaced with the L3+L4-only rule. I.e. you must remove any existing L7 rule associated with the endpoint selectors.

ianvernon replied:
You're right, this does not check whether the current L4Filter has L7 rules involved. We need to perform similar logic to what I have below for cleaning out the L7-related policy.

ianvernon replied:
Fixed and added a unit test.

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -66,6 +71,30 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
+			delete(filter.L7RulesPerEp, k)
+		}
+		filter.L7Parser = ""
+		filter.DerivedFromRules = append(filter.DerivedFromRules, ruleLabels)
```
Here, you're preserving the old rules. I don't think you should, because that won't always be correct.
E.g. if rule A is `allow ingress from app=abc to port 80 HTTP GET /foo`, and rule B is `allow ingress from app=abc to port 80`, there should be no trace of rule A, because it would be incorrect to report that rule A matches a flow with a different URL such as `from app=abc to port 80 HTTP GET /bar`. In this case, I think you should only keep a reference to rule B.

ianvernon replied:
I think the 'DerivedFromRules' sort of gets broken here with the condition you apply. There are multiple cases to consider here:

* Both the rule which we are merging into this L4Filter has no rules *and* the existing L7Parser is empty (no parser); in that case, append the ruleLabels from the rule we are merging into this L4Filter.
* The L4Filter has multiple DerivedFromRules and has a non-empty L7Parser currently (meaning that some rules which were merged into this L4Filter apply at L7), but now we are importing a rule with no L7Rules. What ruleLabels in this case do we remove from the DerivedFromRules list? We can't be sure that these rules are at this location in the code path unless we get the rule itself from the policy repository and analyze it here - at least I think this is the case. What do you think?
* The case you specified, where if we have previously imported rule B, and then import rule A, then we wouldn't add rule A to the DerivedRules because it is shadowed by B.

I view `DerivedFromRules` more of a way of seeing what rules went into the computation of this L4Filter. They could be very well shadowed by other rules; I think it's up to the user to realize that they might have rules which shadow others. I actually think updating `DerivedFromRules` with rules that don't end up doing anything reveals to the user that they have a rule which is functionally useless.

tgraf replied:
As a user I would expect that all rules that contributed to allowing something should be listed. We should refer to `policy trace` for details what gets shadowed where. I would want to know what is affected if I remove a rule. I would look at DerivedRules for that.

rlenglet replied:
OK, makes sense. I was confusing this with `policy trace`.

rlenglet commented on pkg/policy/rule_test.go:
```diff
@@ -1654,3 +1646,105 @@ func (ds *PolicyTestSuite) TestEgressL4AllowAll(c *C) {
+
+	c.Assert(filter.L7Parser, Equals, L7ParserType(""))
+	c.Assert(len(filter.L7RulesPerEp), Equals, 0)
+
```
Please add asserts on `filter.DerivedFromRules`. I'm curious about its contents here.
IMHO it should only refer to the L4-only rule in this case.

ianvernon replied:
I'll punt this for further discussion based off of my comments above. Let's discuss this and nail down what exactly `DerivedFromRules` means.

rlenglet commented on pkg/policy/rule.go:
```diff
@@ -66,21 +77,47 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
+
+	// If rule selects all endpoints and no L7 rules apply, then remove all
+	// L7-related information from the filter and return.
+	if filter.addEndpoints(endpoints) && r.NumRules() == 0 {
```
I'm still not sure this is correct. You should do that only if the new rule selects a superset of the endpoints already selected by the filter.
I'm not sure that `filter.addEndpoints(endpoints)` means `the new rule selects a superset of the filter's endpoints`.

Commit:
Documentation/gettingstarted: update `cilium status` output in Docker GSG

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -447,32 +486,95 @@ func mergeL4Egress(ctx *SearchContext, rule api.EgressRule, ruleLabels labels.La
-			ctx.PolicyTrace("   Merge conflict: mismatching parsers %s/%s\n", l4Filter.L7Parser, filter.L7Parser)
-			return 0, fmt.Errorf("Cannot merge conflicting L7 parsers (%s/%s)", l4Filter.L7Parser, filter.L7Parser)
+
+	// Create a new L4Filter based off of the arguments provided to this function
```
This logic is the same as in mergeL4IngressFilter; this logic should be collapsed together, ideally into a helper function. I can add a TODO for this, but didn't want to spend too much time refactoring due to the short runway to 1.0.

tgraf replied:
This is OK. It was easy to review this way.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
+				existingFilter.L7Parser = filterToMerge.L7Parser
+			}
 		}
+		existingFilter.Endpoints = append(existingFilter.Endpoints, endpoints...)
```
The `append()` reverses the `existingFilter.Endpoints = nil` above. I think this should be in a else branch.

ianvernon replied:
This was actually the logic in `addEndpoints`, which was removed in eb8427a9c3942c8c1a556efdbcbb2f66916a7599 . Per the documentation for `Endpoints` in `L4Filter`: 

https://github.com/cilium/cilium/blob/9571490cbb45bfb93fa230a2c7a6101b03101a56/pkg/policy/l4.go#L88-L90

an empty slice represents all endpoints being selected.

I can test locally and see if a nil slice of EndpointSelector behaves the same as an empty slice.

tgraf replied:
The code will work, it just looks a bit weird to me to set endpoints to nil and then set it to an empty slice again by appending it. I had to read the code twice to make sure. The append is only valid because endpoints is guaranteed to always be an empty slice. I'm OK as-is if you prefer it this way.

ianvernon replied:
Changed because your way makes it more clear; I'm all for more understandable code.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
+	// allows all traffic on L3.
+	//
+	// Existing filter selects all endpoints, so don't add new endpoints.
+	if len(existingFilter.Endpoints) == 0 && len(filterToMerge.Endpoints) > 0 {
```
Is `len(Endpoints) == 0` reliable to determine the wildcard? What if Endpoints has an entry with the `WildcardEndpointSelector`?

ianvernon replied:
`len(Endpoints) == 0` represents all endpoints being selected:  https://github.com/cilium/cilium/blob/9571490cbb45bfb93fa230a2c7a6101b03101a56/pkg/policy/l4.go#L88-L90 

This is confusing since we have the wildcard endpoint selector to represent all endpoints being selected. This should probably be changed to just having the slice contain the WildcardEndpointSelector and then having some helper function for L4Filter called `SelectsAllEndpoints`, which checks if `len(endpoints) == 1` and `endpoints[0] == api.WildcardEndpointSelector`.

ianvernon replied:
Will also add a unit test with the `WildcardEndpointSelector`.

tgraf replied:
If you checked that we are never passing `[]api.EndpointSelector{WildcardEndpointSelector}` in then the code is fine as-is. I agree that we should clean it up in the next feature branch.

ianvernon replied:
Adding explicit allow-all at L3 (i.e., with `WildCardEndpointSelector`) caused breakage with the logic I had in place, so I added a function to check if `EndpointSelectorSlice`selects all endpoints, then changed the list of `Endpoints` in `L4Filter` to be an `EndpointSelectorSlice`. I added new unit tests to validate this behavior, and it now behaves just like the case where if no EndpointSelector was provided at all. Let me know what you think.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
+		// within the filter, and there is no L7 metadata to add to the filter.
+		if r.NumRules() == 0 {
+			existingFilter.DerivedFromRules = append(existingFilter.DerivedFromRules, ruleLabels)
+			return 1, nil
```
Needs to write `existingFilter` back into `resMap` like in other places

ianvernon replied:
We actually don't technically have to here, because the filter already in the map selects all endpoints, and we have no L7-related metadata to update. But, I will do it for consistency's sake.

EDIT: never mind, you're right, we are updating the `DerivedFromRules` :)

ianvernon replied:
Fixed in latest upload.

tgraf reviewed (changes requested):
Awesome work, some small stuff left.

Commit:
pkg/policy/api: add SelectsAllEndpoints function

This function tests whether the EndpointSelectorSlice selects all endpoints.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/k8s: add TODO for cleaning up unit tests

Signed-off by: Ian Vernon <ian@cilium.io>
